File: auth-sdk/typescript/src/settings/opaque.ts
import { AxiosInstance } from "axios";
import * as opaqueLib from "react-native-opaque";
import { OPAQUE_SERVER_IDENTITY } from "../config";
import { TokenManager } from "../token-manager";

export class OpaqueSettings {
  constructor(
    private api: AxiosInstance,
    private tokenManager: TokenManager,
  ) {}

  async setup(password: string) {
    try {
      const token = await this.tokenManager.getAuthToken();
      await opaqueLib.ready;

      const { clientRegistrationState, registrationRequest } =
        opaqueLib.client.startRegistration({ password });

      const challengeRes = await this.api.post(
        `/settings/opaque/setup/challenge`,
        { registrationRequest },
        { headers: { Authorization: `Bearer ${token}` } },
      );

      const registrationResponse = challengeRes.data.registrationResponse;
      if (!registrationResponse)
        throw new Error("Registration response missing");

      const { registrationRecord } = opaqueLib.client.finishRegistration({
        password,
        clientRegistrationState,
        registrationResponse,
        identifiers: { server: OPAQUE_SERVER_IDENTITY },
      });

      const completeRes = await this.api.post(
        `/settings/opaque/setup/complete`,
        { registrationRecord },
        { headers: { Authorization: `Bearer ${token}` } },
      );

      return { success: true, data: completeRes.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || error.message,
      };
    }
  }
}

File: auth-sdk/flutter/lib/settings/opaque.dart
import '../account.dart';
import '../config.dart';
import '../opaque/flutter.dart';
import '../token_manager.dart';

class OpaqueSettings {
  OpaqueSettings(this.api, this.tokens);
  final AuthApi api;
  final TokenManager tokens;
  final opaque = FlutterOpaqueClient();
  Future<AuthResult<JsonMap>> setup(String password) async {
    try {
      final start = await opaque.startRegistration(password);
      final jwt = await tokens.get();
      final challenge = await api.send(
          'POST', '/settings/opaque/setup/challenge',
          body: {'registrationRequest': start.registrationRequest},
          headers: authHeaders(jwt));
      final record = await opaque.finishRegistration(
          password: password,
          clientRegistrationState: start.clientRegistrationState,
          registrationResponse: challenge['registrationResponse'] as String,
          serverIdentity: opaqueServerIdentity);
      return AuthResult.success(await api.send(
          'POST', '/settings/opaque/setup/complete',
          body: {'registrationRecord': record}, headers: authHeaders(jwt)));
    } catch (e) {
      return AuthResult.failure('$e');
    }
  }
}
